<a href="https://colab.research.google.com/github/syulimo/3rd-ML100Days/blob/master/Intro%20Notebook%20-%20Key.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Intro Notebook
### Rice Data Science - DEEP Team 3
#### [Insert Name Here]

This notebook is designed as an environment to get an intro to doing data wrangling with pandas as well as starting to get insights into the thyroid data set. To begin we need to load the necessary packages.

In [ ]:
# Loading in packages
import pandas as pd
import numpy as np

Next we want to import the thyroid csv file into the notebook. One way to do this in colab is to first download the data set on to your computer and then upload the file into colab by clicking the folder icon on the left hand side of screen under the key and search icons. Once you open this tab, directly under files there should be an upload icon that allows you to upload the file for use in the Colab notebook. Now we can load the csv file into our notebook using pd.read_csv

In [ ]:
# CODE HERE: Read thyroid data into notebook using pandas and assign it to a variable named `thyroid_dat`
thyroid_dat = pd.read_csv("thyroidDF.csv")

### Part 1: First Steps When Working on New Data

In any data science project, there's a few things you will want to focus on once you have got your data together in a usable format:
1. How much data do you have?
2. What does your data look like?
3. What types of data do you have?
4. How much data is missing (and why is it missing)?
5. How is your data distributed?

Some of these are more open ended than others and some will require tools we will learn in later workshops, but we can start by answers a couple of them using simple methods.

Let's start with figuring out how much data we are working with; `based on what we learned from the previous workshop, what method/attribute should we use?`

In [ ]:
# CODE HERE: Print the number of rows and columns in `thyroid_dat`
thyroid_dat.shape

(9172, 31)

I would consider this to be a moderate amount of data; not in the millions or billions of observations to be considered "big data" (where you have too much data to load into memory), but also not a few hundred rows where you need to be much more careful about the statistical validity of the methods you use. In terms of features, I would also say this is a moderate to high amount; it will be somewhat difficult to completely understand 31 different variables, but it is not on the scale of having hundreds or more features where you likely would need to find workarounds to comprehend all the data you are working with. Overall, I would say this is quite a manageble data set so far.

Next we would like to get some sense of what the data actually looks like; `what approaches were covered in the workshop to directly look at our data?`

In [ ]:
# CODE HERE: Take a look at `thyroid_dat` using pandas
thyroid_dat.head()

age sex on_thyroxine query_on_thyroxine on_antithyroid_meds sick pregnant  \
0   29   F            f                  f                   f    f        f   
1   29   F            f                  f                   f    f        f   
2   41   F            f                  f                   f    f        f   
3   36   F            f                  f                   f    f        f   
4   32   F            f                  f                   f    f        f   

  thyroid_surgery I131_treatment query_hypothyroid  ...    TT4 T4U_measured  \
0               f              f                 t  ...    NaN            f   
1               f              f                 f  ...  128.0            f   
2               f              f                 f  ...    NaN            f   
3               f              f                 f  ...    NaN            f   
4               f              f                 f  ...    NaN            f   

  T4U FTI_measured FTI TBG_measured   TBG  referral_source target  patient_id  
0 NaN            f NaN            f   NaN            other      -   840801013  
1 NaN            f NaN            f   NaN            other      -   840801014  
2 NaN            f NaN            t  11.0            other      -   840801042  
3 NaN            f NaN            t  26.0            other      -   840803046  
4 NaN            f NaN            t  36.0            other      S   840803047  

[5 rows x 31 columns]

In [ ]:
thyroid_dat.tail()

age sex on_thyroxine query_on_thyroxine on_antithyroid_meds sick  \
9167   56   M            f                  f                   f    f   
9168   22   M            f                  f                   f    f   
9169   69   M            f                  f                   f    f   
9170   47   F            f                  f                   f    f   
9171   31   M            f                  f                   f    f   

     pregnant thyroid_surgery I131_treatment query_hypothyroid  ...    TT4  \
9167        f               f              f                 f  ...   64.0   
9168        f               f              f                 f  ...   91.0   
9169        f               f              f                 f  ...  113.0   
9170        f               f              f                 f  ...   75.0   
9171        f               f              f                 t  ...   66.0   

     T4U_measured   T4U FTI_measured   FTI TBG_measured TBG  referral_source  \
9167            t  0.83            t  77.0            f NaN              SVI   
9168            t  0.92            t  99.0            f NaN              SVI   
9169            t  1.27            t  89.0            f NaN              SVI   
9170            t  0.85            t  88.0            f NaN            other   
9171            t  1.02            t  65.0            f NaN            other   

     target  patient_id  
9167      -   870119022  
9168      -   870119023  
9169      I   870119025  
9170      -   870119027  
9171      -   870119035  

[5 rows x 31 columns]

We already took some time to look at this data set and state some things that were notable about it. One thing mentioned was that the data include multiple data types. Pandas makes it extremely easy to summarize what data types we have in a couple of ways. `How do we display the data types of every column of a data set?`

In [ ]:
# CODE HERE: Print the data types of every column in the `thyroid_dat` data frame
thyroid_dat.dtypes

age                      int64
sex                     object
on_thyroxine            object
query_on_thyroxine      object
on_antithyroid_meds     object
sick                    object
pregnant                object
thyroid_surgery         object
I131_treatment          object
query_hypothyroid       object
query_hyperthyroid      object
lithium                 object
goitre                  object
tumor                   object
hypopituitary           object
psych                   object
TSH_measured            object
TSH                    float64
T3_measured             object
T3                     float64
TT4_measured            object
TT4                    float64
T4U_measured            object
T4U                    float64
FTI_measured            object
FTI                    float64
TBG_measured            object
TBG                    float64
referral_source         object
target                  object
patient_id               int64
dtype: object

What you will notice is that we have a mix of 64 bit integers, 64 bit floats, and object data types in this data set. Note that an object data type can be thought of as a general data type, although in most cases it indicates character or string data. `Do you notice any differences between the columns with an object data type?`

To better understand the differences between these columns we can check how many unique entries there are in each object column. `What attribute/method would be useful to do this?`

In [ ]:
# CODE HERE: 1. Filter the data set to only include columns with an object data type 2. Check how many unique entries are in these columns
thyroid_dat.select_dtypes(include = "object").nunique()

sex                     2
on_thyroxine            2
query_on_thyroxine      2
on_antithyroid_meds     2
sick                    2
pregnant                2
thyroid_surgery         2
I131_treatment          2
query_hypothyroid       2
query_hyperthyroid      2
lithium                 2
goitre                  2
tumor                   2
hypopituitary           2
psych                   2
TSH_measured            2
T3_measured             2
TT4_measured            2
T4U_measured            2
FTI_measured            2
TBG_measured            2
referral_source         6
target                 32
dtype: int64

Next, we would like to take a look at missing data. You will notice when we used the `.head()` method there were some `NaN` values, and we would like to better understand how much missing data we are working with before going much further.

In [ ]:
# CODE HERE: Print out the number of missing rows per column (consider using `.isna()`)
thyroid_dat.isna().sum()

age                       0
sex                     307
on_thyroxine              0
query_on_thyroxine        0
on_antithyroid_meds       0
sick                      0
pregnant                  0
thyroid_surgery           0
I131_treatment            0
query_hypothyroid         0
query_hyperthyroid        0
lithium                   0
goitre                    0
tumor                     0
hypopituitary             0
psych                     0
TSH_measured              0
TSH                     842
T3_measured               0
T3                     2604
TT4_measured              0
TT4                     442
T4U_measured              0
T4U                     809
FTI_measured              0
FTI                     802
TBG_measured              0
TBG                    8823
referral_source           0
target                    0
patient_id                0
dtype: int64

Unfortunately, we have enough columns that the output gets truncated, so let's try again filtering for columns that have at least one missing value.

In [ ]:
# Number of missing rows by column
print(thyroid_dat.loc[:, thyroid_dat.isna().sum() > 1].isna().sum(), "\n")

# Proportion of missing rows to total rows
print(thyroid_dat.loc[:, thyroid_dat.isna().sum() > 1].isna().sum() / thyroid_dat.shape[0] * 100)

sex     307
TSH     842
T3     2604
TT4     442
T4U     809
FTI     802
TBG    8823
dtype: int64 

sex     3.347143
TSH     9.180113
T3     28.390754
TT4     4.819014
T4U     8.820323
FTI     8.744003
TBG    96.194941
dtype: float64


From this we can see that the `TBG` variable is missing for more than 96% of the data, and the `T3` variable is missing for more than a quarter of the observations. Now before we start trying to deal with this missing data we should first ask ourselves, `what reasons could there be for this data to be missing?`

1. Structural Missingness: Data is missing because it should be missing. \
**Example:** Data about types of meat consumed is missing for a herbivore.
2. Missing Completely at Random (MCAR): Data is missing for no reason. \
**Example:** You take a data set and replace 10% of the data with missing values.
3. Missing at Random (MAR): Data is more likely to be missing due to predictable relationship. \
**Example:** People are less likely to answer a question about using ChatGPT for homework if they are under the age of 22.
4. Missing not at Random (MNAR): Data is missing for unclear reasons. \
**Example:** Economic data about wheat production is missing for Sweden, Kazakhstan, Australia, and South Africa.

Note that structural missingness is a special case of MNAR, however the main difference is that while structural missingness is easy to explain and often to model, the vast majority of data sets with data that is MNAR can run into massive issues when drawing generalizable conclusions.

As you might be able to tell, figuring out why data is missing can be an analysis project on its own depending on the complexity of the data, and for now we will wait to dive deeper in our analysis until we have more tools to work with.

The final major thing we want to get a handle on is how our data is distributed. For now, we will stick to numerical summaries of the data, but you could also think of things like histograms and box plots as being useful tools for this. `What Pandas methods can you use to get an idea of the distribution of the data?`

In [ ]:
# CODE HERE: Get a numerical summary of each numerical variable in the `thyroid_dat` data frame
thyroid_dat.describe()

age          TSH           T3          TT4          T4U  \
count   9172.000000  8330.000000  6568.000000  8730.000000  8363.000000   
mean      73.555822     5.218403     1.970629   108.700305     0.976056   
std     1183.976718    24.184006     0.887579    37.522670     0.200360   
min        1.000000     0.005000     0.050000     2.000000     0.170000   
25%       37.000000     0.460000     1.500000    87.000000     0.860000   
50%       55.000000     1.400000     1.900000   104.000000     0.960000   
75%       68.000000     2.700000     2.300000   126.000000     1.065000   
max    65526.000000   530.000000    18.000000   600.000000     2.330000   

               FTI         TBG    patient_id  
count  8370.000000  349.000000  9.172000e+03  
mean    113.640746   29.870057  8.529473e+08  
std      41.551650   21.080504  7.581969e+06  
min       1.400000    0.100000  8.408010e+08  
25%      93.000000   21.000000  8.504090e+08  
50%     109.000000   26.000000  8.510040e+08  
75%     128.000000   31.000000  8.607110e+08  
max     881.000000  200.000000  8.701190e+08

While without good domain knowledge it is hard to get much insight from a lot of these summarys, it is clear that the numerical columns have fairly different scales, which we may need to account for going forward. We can also notice that there is something odd with the `age` column as the maximum age is apparently 65526 which obviously does not make much sense. When we get around to cleaning the data this is something to keep in mind.

However we don't want to be limited to just understanding the distribution of the numerical variables, and thankfully we can fairly easily describe categorical columns by just finding the number of observations by class. `What Pandas method counts the number of observations by value?`

In [ ]:
# Find distribution of boolean variables
thyroid_dat.loc[:, thyroid_dat.nunique() == 2].apply(pd.Series.value_counts)

sex  on_thyroxine  query_on_thyroxine  on_antithyroid_meds    sick  \
F  6073.0           NaN                 NaN                  NaN     NaN   
M  2792.0           NaN                 NaN                  NaN     NaN   
f     NaN        7932.0              9019.0               9056.0  8828.0   
t     NaN        1240.0               153.0                116.0   344.0   

   pregnant  thyroid_surgery  I131_treatment  query_hypothyroid  \
F       NaN              NaN             NaN                NaN   
M       NaN              NaN             NaN                NaN   
f    9065.0           9038.0          9003.0             8542.0   
t     107.0            134.0           169.0              630.0   

   query_hyperthyroid  ...  goitre   tumor  hypopituitary   psych  \
F                 NaN  ...     NaN     NaN            NaN     NaN   
M                 NaN  ...     NaN     NaN            NaN     NaN   
f              8521.0  ...  9088.0  8931.0         9170.0  8754.0   
t               651.0  ...    84.0   241.0            2.0   418.0   

   TSH_measured  T3_measured  TT4_measured  T4U_measured  FTI_measured  \
F           NaN          NaN           NaN           NaN           NaN   
M           NaN          NaN           NaN           NaN           NaN   
f         842.0       2604.0         442.0         809.0         802.0   
t        8330.0       6568.0        8730.0        8363.0        8370.0   

   TBG_measured  
F           NaN  
M           NaN  
f        8823.0  
t         349.0  

[4 rows x 21 columns]

In [ ]:
# Find distribution of target variable
thyroid_dat["target"].value_counts()

age sex on_thyroxine query_on_thyroxine on_antithyroid_meds sick  \
2976    455   F            f                  f                   f    f   
5710  65511   M            f                  f                   f    f   
6392  65512   M            f                  f                   f    f   
8105  65526   F            f                  f                   f    f   

     pregnant thyroid_surgery I131_treatment query_hypothyroid  ...    TT4  \
2976        f               f              f                 f  ...  118.0   
5710        f               f              f                 f  ...  113.0   
6392        f               f              f                 f  ...  112.0   
8105        f               f              f                 f  ...  132.0   

     T4U_measured   T4U FTI_measured    FTI TBG_measured TBG  referral_source  \
2976            t  1.13            t  104.0            f NaN              SVI   
5710            t  1.08            t  104.0            f NaN            other   
6392            t  0.84            t  133.0            f NaN            other   
8105            t  1.02            t  129.0            f NaN            other   

     target  patient_id  
2976      -   850530001  
5710      -   860210008  
6392      -   860403050  
8105      -   861014041  

[4 rows x 31 columns]

Naturally you can do through each categorical variable and call `value_counts()` on it, although we are just aiming to get a general idea of how our data is structured, and we can take a closer look later on.

### Part 2: Data Cleaning

Once we have a basic grasp of our data through the process above we are ready to start doing data cleaning. Data cleaning comprises of steps taken to make a data set more usable whether this be handling missing data, fixing incorrect data, etc. Data cleaning and wrangling is almost always the majority of time you will spend when doing a data science project and as such it can get quite complicated depending on the data you are working with. Thankfully our data is quite straightforward and high quality, so we will only need to take a few steps to get it into good shape.

For starters, let's take a look at that weird issues we found with age, where the maximum age was 65,526. Let's look at all the rows where age is higher than 100 to see if we can spot unreasonable outliers.

In [ ]:
# CODE HERE: Print all the rows of `thyroid_dat` where age is greater than 100
thyroid_dat[thyroid_dat["age"] > 100]

age sex on_thyroxine query_on_thyroxine on_antithyroid_meds sick  \
2976    455   F            f                  f                   f    f   
5710  65511   M            f                  f                   f    f   
6392  65512   M            f                  f                   f    f   
8105  65526   F            f                  f                   f    f   

     pregnant thyroid_surgery I131_treatment query_hypothyroid  ...    TT4  \
2976        f               f              f                 f  ...  118.0   
5710        f               f              f                 f  ...  113.0   
6392        f               f              f                 f  ...  112.0   
8105        f               f              f                 f  ...  132.0   

     T4U_measured   T4U FTI_measured    FTI TBG_measured TBG  referral_source  \
2976            t  1.13            t  104.0            f NaN              SVI   
5710            t  1.08            t  104.0            f NaN            other   
6392            t  0.84            t  133.0            f NaN            other   
8105            t  1.02            t  129.0            f NaN            other   

     target  patient_id  
2976      -   850530001  
5710      -   860210008  
6392      -   860403050  
8105      -   861014041  

[4 rows x 31 columns]

From this we can see that there is 4 rows where a patient has an obviously incorrect age. There is a few ways we could handle this. First, we have enough data that we could probably get away with just dropping these columns. However, if we don't want to do this then instead we could change these ages to missing values and deal with them when we deal with missing values.

In [ ]:
# Replace any values of age that are greater than 100 with NaN
thyroid_dat['age'] = thyroid_dat['age'].apply(lambda x: np.NaN if x > 100 else x)

It seems as though all of our columns are the correct type (i.e. no numerical columns are encoded as strings), so we can move on to dealing with the missing data. There are many different ways to deal with missing values; in the kickoff meeting you all mentioned things like dropping rows with missing values, imputing missing values with the mean, etc. Let's go over the pros and cons of some of these methods.

#### 1. Dropping Rows with Missing Vlaues
**Pros:** Quick and easy to perform; don't need to worry about changing distribution of data. \
**Cons:** Easy to drop too much data; dropped rows may contain valuable information; tacitly assumes data is MCAR. \
**Use Cases:**
1. Data sets with relatively small amounts of missing data.
2. Data sets where some columns are missing data in most or all columns.
3. Data sets with rows missing the response/target variable.

**Notes:** \
Dropping rows is one of the most rudimentary ways to deal with missing values, and while it does have legitimate use cases you do need to be very careful and ensure that dropping all rows with even one missing value does not lead to you dropping 96% of your dataset like it would in our case. Even more important is to ensure that dropping rows, even if it is just a few, is not resulting in you losing out on information in the dataset.


In [ ]:
# CODE HERE: Drop every row in the data set that has a missing value (do NOT assign it to `thyroid_dat` and ensure inplace=False)
thyroid_dat.dropna(inplace=False)

age sex on_thyroxine query_on_thyroxine on_antithyroid_meds sick  \
167   40.0   F            f                  f                   f    f   
5256  35.0   F            f                  f                   f    f   
6044  77.0   F            f                  f                   f    f   
6045  73.0   M            f                  f                   f    f   
6747  77.0   F            f                  f                   f    f   
6773  74.0   F            f                  f                   f    f   
6862  60.0   M            f                  f                   f    f   
6863  66.0   F            f                  f                   f    f   
6880  42.0   F            f                  f                   f    f   
6934  29.0   F            f                  f                   f    f   
7097  35.0   M            f                  f                   f    f   
7101  44.0   F            f                  f                   f    f   
7103  78.0   F            f                  f                   f    f   
7190  28.0   F            f                  f                   f    f   
7239  58.0   M            f                  f                   f    f   
7254  46.0   F            f                  f                   f    f   
7356  97.0   M            f                  f                   f    f   
7591  57.0   F            f                  f                   f    f   
7733  60.0   F            f                  f                   f    f   
8116  24.0   F            t                  f                   f    f   

     pregnant thyroid_surgery I131_treatment query_hypothyroid  ...    TT4  \
167         f               f              f                 f  ...    3.9   
5256        f               t              f                 f  ...   73.0   
6044        f               f              f                 f  ...  120.0   
6045        f               f              f                 f  ...   89.0   
6747        f               f              f                 f  ...  131.0   
6773        f               f              f                 f  ...  116.0   
6862        f               f              f                 f  ...   92.0   
6863        f               f              f                 f  ...  138.0   
6880        f               f              f                 f  ...  106.0   
6934        f               f              f                 f  ...  122.0   
7097        f               f              f                 f  ...   97.0   
7101        f               f              f                 f  ...  135.0   
7103        f               f              f                 f  ...  120.0   
7190        f               f              f                 f  ...   75.0   
7239        f               f              f                 f  ...  161.0   
7254        f               f              f                 f  ...  116.0   
7356        f               f              f                 f  ...   93.0   
7591        f               f              f                 f  ...  110.0   
7733        f               f              f                 f  ...  139.0   
8116        f               f              f                 f  ...  176.0   

     T4U_measured   T4U FTI_measured    FTI TBG_measured   TBG  \
167             t  0.83            t    5.0            t  28.0   
5256            t  1.16            t   63.0            t  37.0   
6044            t  0.96            t  124.0            t  45.0   
6045            t  0.74            t  119.0            t  24.0   
6747            t  1.04            t  126.0            t  25.0   
6773            t  0.81            t  143.0            t  22.0   
6862            t  0.84            t  110.0            t  21.0   
6863            t  0.80            t  173.0            t  15.0   
6880            t  0.98            t  108.0            t  27.0   
6934            t  1.14            t  107.0            t  36.0   
7097            t  0.85            t  114.0            t  30.0   



#### 2. Replacing Missing Values with the Mean (or Median or Mode)
**Pros:** Quick and easy to perform; let's you retain all the rows in your data set. \
**Cons:** Reduces variance of the variable; imputation may not make logical sense. \
**Use Cases:**
1. Data sets with relatively small amounts of missing data.
2. Very large data sets where other techniques are too computationally intensive.

**Notes:** \
Imputing a single value for all missing values in a column is also very rudimentary and can get you in a lot of trouble in some cases. Ultimately, the thing to keep in mind is that imputing a single value is not simply going to replace the missing values, it is also going to change the data itself and the relationships within it. This can result in whatever analysis or modeling you do on this data to be misleading and make wrong conclusions because the data you are working with is not representative anymore of reality.


In [ ]:
# CODE HERE: 1. Impute the age column using the mean
thyroid_dat['age'] = thyroid_dat['age'].fillna(thyroid_dat['age'].mean(), inplace=False)

print('Number of null age values: ', thyroid_dat.isna().sum()['age']) # Code to check how many entries in age are null

Number of null age values:  0



#### 3. Replacing Missing Values with a Randomly Sampled Value
**Pros:** Fairly quick and easy to perform; let's you retain all the rows in your data set; typically will not change distribution of data. \
**Cons:** Imputation may not make logical sense; effectively just adding noise to the data. \
**Use Cases:**
1. Data sets with relatively small amounts of missing data.
2. Very large data sets where other techniques are too computationally intensive.

**Notes:** \
Replacing each missing value in a column with a randomly picked value already within the column can be an quick imputation strategy that does not have the con of mean imputation where you are changing the distribution of the data. However, the fact that you are randomly picking the value means that it inherently holds no actual information that would be useful for analysis or modeling. This can reduce predictive performance and make trends and other relationships harder to spot in analysis.


In [ ]:
# Example of how to impute using a randomly sampled value
non_missing_values = thyroid_dat['sex'].dropna()
thyroid_dat['sex'] = thyroid_dat['sex'].apply(lambda x: np.random.choice(non_missing_values) if pd.isnull(x) else x)
print('Number of null sex values: ', thyroid_dat.isna().sum()['sex'])

Number of null sex values:  0



#### 4. Replacing Missing Values using Machine Learning
**Pros:** In the best case can replace missing values with actually meaningful values. \
**Cons:** Much more time-intensive even with simple models; no guarantee that imputations are useful. \
**Use Cases:**
1. Data sets with medium to large amounts of missing data in some columns.
2. Data sets where there are clear relationships between variables and missingness.
3. Data sets where simple techniques produce poor results.
4. Data sets with correlated predictor variables.

**Notes:** \
Using a machine learning model to impute missing values is a much more accurate way to do imputation than methods discussed so far. It is typically done using non-parametric models like k-means or random forest. Even for a simple model like k-means, it will take much longer to perform compared to simple techniques mentioned above. This also means you will need to deal with another model which can increase the complexity of a project substantially. While I have mentioned quite a few drawbacks, I would say in most cases you should use a machine learning based approach as it should reduce potential headaches caused by the other methods that only show up at the end of the project.


In [ ]:
# Example of doing machine learning imputation using K Nearest Neighbors
# Note K-NN imputation only works for numeric data
from sklearn.impute import KNNImputer

# Initialize KNNImputer with a specified number of neighbors (default is 5)
imputer = KNNImputer(n_neighbors=3)

# Apply the KNN imputer to the DataFrame
df_imputed = pd.DataFrame(imputer.fit_transform(thyroid_dat.select_dtypes(include = ["int", "float"])), columns=thyroid_dat.select_dtypes(include = ["int", "float"]).columns)
df_imputed.isna().sum()

age           0
TSH           0
T3            0
TT4           0
T4U           0
FTI           0
TBG           0
patient_id    0
dtype: int64


#### 5. Multiple Imputation
**Pros:** Can greatly reduce random effects of a single imputation on results. \
**Cons:** Multiplies the amount of computational work of all subsequent data pipeline steps; much harder to do data analysis on multiple data sets concurrently than one. \
**Use Cases:**
1. Technically can be applied to any of the previous methods, but is more useful for non-deterministic methods like random sampling and machine learning.
2. Especially useful for data sets missing a lot of data.
3. When you need to quantify the amount of uncertainty in missing data.

**Notes:** \
Multiple imputation can be an effective imputation strategy if it appears as though the results of your analysis/modeling are highly affected by imputed values. In my experience it is much more time consuming to be doing analysis on multiple slightly different data sets, and you will want to have good helper functions to reduce the amount of manual work you need to do when training models on each data set and combining results. One really useful application of multiple imputation is that it allows you to understand how much uncertainty there is in the missing data which can be critical when doing statistical inference.

We won't dive into how to do multiple imputation as it takes a fair bit of set up, but it is a good thing to think about when doing future data projects.

#### Numeric Encoding

Most models require all input data to be numeric, so we need to encode all of the boolean and string variables into a numeric format. This is fairly simple for boolean variables as we just need to set one class to 0 and the other to 1. It gets much more complicated for string variables especially free text, but we will not need to worry about that for this data set. What we do need to worry about is string variables that represents multiple classes. Before that let's convert all our boolean variables to 0's and 1's.

In [ ]:
# CODE HERE: Encode boolean variables as 0's and 1's (there is a single pandas function that allows you to do this)
# Note you need to apply the transformation on only the boolean columns
pd.get_dummies(thyroid_dat, columns=thyroid_dat.loc[:, thyroid_dat.nunique() == 2].columns, drop_first=True)

age  TSH   T3    TT4   T4U   FTI   TBG referral_source target  \
0     29.0  0.3  NaN    NaN   NaN   NaN   NaN           other      -   
1     29.0  1.6  1.9  128.0   NaN   NaN   NaN           other      -   
2     41.0  NaN  NaN    NaN   NaN   NaN  11.0           other      -   
3     36.0  NaN  NaN    NaN   NaN   NaN  26.0           other      -   
4     32.0  NaN  NaN    NaN   NaN   NaN  36.0           other      S   
...    ...  ...  ...    ...   ...   ...   ...             ...    ...   
9167  56.0  NaN  NaN   64.0  0.83  77.0   NaN             SVI      -   
9168  22.0  NaN  NaN   91.0  0.92  99.0   NaN             SVI      -   
9169  69.0  NaN  NaN  113.0  1.27  89.0   NaN             SVI      I   
9170  47.0  NaN  NaN   75.0  0.85  88.0   NaN           other      -   
9171  31.0  NaN  NaN   66.0  1.02  65.0   NaN           other      -   

      patient_id  ...  goitre_t  tumor_t  hypopituitary_t  psych_t  \
0      840801013  ...     False    False            False    False   
1      840801014  ...     False    False            False    False   
2      840801042  ...     False    False            False    False   
3      840803046  ...     False    False            False    False   
4      840803047  ...     False    False            False    False   
...          ...  ...       ...      ...              ...      ...   
9167   870119022  ...     False    False            False    False   
9168   870119023  ...     False    False            False    False   
9169   870119025  ...     False    False            False    False   
9170   870119027  ...     False    False            False    False   
9171   870119035  ...     False    False            False    False   

      TSH_measured_t  T3_measured_t  TT4_measured_t  T4U_measured_t  \
0               True          False           False           False   
1               True           True            True           False   
2              False          False           False           False   
3              False          False           False           False   
4              False          False           False           False   
...              ...            ...             ...             ...   
9167           False          False            True            True   
9168           False          False            True            True   
9169           False          False            True            True   
9170           False          False            True            True   
9171           False          False            True            True   

      FTI_measured_t  TBG_measured_t  
0              False           False  
1              False           False  
2              False            True  
3              False            True  
4              False            True  
...              ...             ...  
9167            True           False  
9168            True           False  
9169            True           False  
9170            True           False  
9171            True           False  

[9172 rows x 31 columns]

Next we want to deal with the categorical columns that have multiple classes within them rather than just two. To do this we first need to identify which columns these are.

In [ ]:
# CODE HERE: Which columns are categorical and have more than 2 categories?
thyroid_dat.loc[:, (thyroid_dat.dtypes == "object") & (thyroid_dat.nunique() > 2)].nunique()

referral_source     6
target             32
dtype: int64

We have two multi-class categorical columns: `referral_source` with 6 different classes, and `target` with 32 different classes. Given that there are relatively few classes in `referral_source` we can make use of **one-hot encoding** where we create 5 columns for 5 of the classes, and each column is an indicator of whether that row has that particular class. One-hot encoding is a more general case of the encoding we did with the boolean variables, so we can reuse the same pandas function for it.

In [ ]:
# CODE HERE: Encode referral_source using one-hot encoding
pd.get_dummies(thyroid_dat, columns="referral_source")

Now for the `target` variable we need to be much more careful as this is what we want to predict. To get a better idea about how to deal with this variable let's take a look at what the classes actually are.

In [ ]:
# CODE HERE: Print the unique values of target
thyroid_dat["target"].unique()

array(['-', 'S', 'F', 'AK', 'R', 'I', 'M', 'N', 'G', 'K', 'A', 'KJ', 'L',
       'MK', 'Q', 'J', 'C|I', 'O', 'LJ', 'H|K', 'D', 'GK', 'MI', 'P',
       'FK', 'B', 'GI', 'C', 'GKJ', 'OI', 'D|R', 'E'], dtype=object)

What you will notice is that alot of these are combinations of multiple letters. Another important thing to notice is that some of these classes have a line between them. Take a look at the data set information document; `what does this tell us about this patient?`

### Part 3: Encoding Target Variable

This will be a more open-ended section as there is no go to approach for encoding the `target` variable properly. Think about how you can encode all the inherent information between the different classes while still retaining the fact that some combinations represent multiple conditions and others represent conditions that are not completely clear.

In [ ]:
# CODE HERE: Encode target variable as numbers